### PR TITLE
fix: JSON serialization of `metrics::Examplar` and `trace::span::Link…

### LINKED ISSUE #2065: [Bug]: Invalid link serialization in HTTP/JSON exporter
### What happened?

When using the HTTP/JSON exporter, JSON serialization of links is invalid.

Here is an example JSON generated for a span with a link:

```json
            {
              "traceId": "e77be372af2941b6b25c2a2462f54f5e",
              "spanId": "7285b8247c8441cd",
              "traceState": "",
              "parentSpanId": "a9acb8895a774ede",
              "flags": 1,
              "name": "AuthService::grant_oauth_authorization::claims",
              "kind": 1,
              "startTimeUnixNano": "1724896750141000000",
              "endTimeUnixNano": "1724896750141000000",
              "attributes": [],
              "droppedAttributesCount": 0,
              "events": [],
              "droppedEventsCount": 0,
              "links": [
                {
                  "traceId": [
                    157,
                    48,
                    54,
                    48,
                    139,
                    160,
                    71,
                    67,
                    152,
                    135,
                    40,
                    155,
                    119,
                    66,
                    113,
                    189
                  ],
                  "spanId": [
                    233,
                    128,
                    29,
                    199,
                    119,
                    206,
                    74,
                    251
                  ],
                  "traceState": "",
                  "attributes": [],
                  "droppedAttributesCount": 0,
                  "flags": 1
                }
              ],
              "droppedLinksCount": 0,
              "status": {
                "message": "",
                "code": 0
              }
            },
```

As you see, the fields `traceId` and `spanId` are written as arrays of ints. However [version `1.3.2` of the OTLP spec](https://opentelemetry.io/docs/specs/otlp/#json-protobuf-encoding) requires hex for these field:

```
> The traceId and spanId byte arrays are represented as [case-insensitive hex-encoded strings](https://tools.ietf.org/html/rfc4648#section-8); they are not base64-encoded as is defined in the standard [Protobuf JSON Mapping](https://developers.google.com/protocol-buffers/docs/proto3#json). Hex encoding is used for traceId and spanId fields in all OTLP Protobuf messages, e.g., the Span, Link, LogRecord, etc. messages. For example, the traceId field in a Span can be represented like this: { “traceId”: “5B8EFFF798038103D269B633813FC60C”, … }
```

### API Version

0.24.0

### SDK Version

0.24.1

### What Exporter(s) are you seeing the problem on?

OTLP

### Relevant log output

```shell
OpenTelemetry trace error occurred. client error (SendRequest)
```

**EDIT**: Also applies to `metrics.Exemplar`

### PR DESCRIPTION
**Pull request by demurgos**
* [x] add more tests

Fixes #2065
Design discussion issue (if applicable): N/A

## Changes

This commit fixes open-telemetry/opentelemetry-rust#2065: it uses string serialization for all `TraceId` and `SpanId` fields. In particular, this fixes serialization of `metrics::Examplar` and `trace::span::Link`.

As part of the change, JSON deserialization tests were greatly expanded to test both serialization and deserialization, with support for alternative serializations. The tests are extended to cover all examples from the `opentelemetry-proto` repository, it also adds a test where all values are used (without any default primitives or empty array).

## Merge requirement checklist

* [x] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-rust/blob/main/CONTRIBUTING.md) guidelines followed
* [x] Unit tests added/updated (if applicable)
* [x] Appropriate `CHANGELOG.md` files updated for non-trivial, user-facing changes
* [x] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: opentelemetry-proto/CHANGELOG.md [hunk 1]
```diff
@@ -1,7 +1,9 @@
 # Changelog
 
 ## vNext
--  Update protobuf definitions to v1.3.2 [#1945](https://github.com/open-telemetry/opentelemetry-rust/pull/1945)
+
+- Update protobuf definitions to v1.3.2 [#1945](https://github.com/open-telemetry/opentelemetry-rust/pull/1945)
+- Fix JSON serialization of `metrics::Exemplar` and `trace::span::Link`
```
Comment (TommyCpp) [inline opentelemetry-proto/CHANGELOG.md:6]: nit: could you link the PR here

**Comment by lalitb** [comment]
Thanks @demurgos, this is quite extensive. And also good to have the validation for examples from otel-proto. Couple of nit comments, apart from the CI lint error to be taken care of.

**Comment by demurgos** [comment]
Thank you for the feedback, I'll fix the typos and lint errors.

File: opentelemetry-proto/CHANGELOG.md (hunk 1 above)
Comment (demurgos) [reply opentelemetry-proto/CHANGELOG.md]: Fixed: I added the PR link.

File: opentelemetry-proto/tests/grpc_build.rs [hunk 2]
```diff
@@ -41,8 +41,8 @@ fn build_tonic() {
             "#[cfg_attr(feature = \"with-serde\", serde(rename_all = \"camelCase\"))]",
         );
 
-    // optional numeric and String field need to default it to 0 otherwise JSON files without those field
-    // cannot deserialize
+    // optional numeric, string and array fields need to default it their default value otherwise
```
[the remark this answers is not in the document]
Comment (demurgos) [reply opentelemetry-proto/tests/grpc_build.rs]: Fixed

File: opentelemetry-proto/tests/json_serde.rs [hunk 3]
```diff
@@ -0,0 +1,1482 @@
...
+
+      #[test]
+      fn deserialize_alternative() {
+        let actual: ExportTraceServiceRequest = serde_json::from_str(ALTERNATIVE).expect("deserialization must succeed");
+        let expected: ExportTraceServiceRequest = value();
+        assert_eq!(actual, expected);
+      }
+    }
+
+    // request were all fields are set
```
[the remark this answers is not in the document]
Comment (demurgos) [reply opentelemetry-proto/tests/json_serde.rs]: Fixed

**Comment by demurgos** [comment]
I fixed the typos, changelog and formatting.

I also fixed the linter error. Some of them were due to feature combinations so I had to add some extra `cfg(...)` in the tests. I also update the linter script to check for the relevant feature combinations (for `logs` and `metrics`).

**Review by lalitb — APPROVED** [review]

**Comment by lalitb** [comment]
> The remaining error are spurious diffs in the generated protobuf code. I'm not sure if it's just a bit flaky or related to my changes.

This is fixed in #2077. CI should hopefully pass once the conflict is resolved.

**Comment by demurgos** [comment]
@lalitb I rebased my PR on top of main. CI is now green.
